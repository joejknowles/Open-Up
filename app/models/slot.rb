class Slot < ApplicationRecord
  validates :start_time, presence: true
end
