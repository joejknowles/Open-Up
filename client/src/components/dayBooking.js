import React, { Component } from 'react';
import '../styles/App.css';
import { connect } from 'react-redux';

export class DayBooking extends Component {
  render() {
    const date = this.props.date ?
      this.props.date.toLocaleDateString("en-GB") :
      '';
    const slots = this.props.slots || []
    return (
      <div>
        <h3>Today { date }</h3>
        { slots.map(({ id }) => (
            <button key={ id } ></button>
        )) }
      </div>
    );
  }
}

const mapStateToProps = (state) => ({
  slots: state.allSlots.map((id) => state.slotsById[id]),
  date: state.selectedDate
});

export default connect(mapStateToProps)(DayBooking);
