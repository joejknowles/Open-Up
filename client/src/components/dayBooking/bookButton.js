import React from 'react';
import { connect } from 'react-redux';
import format from 'date-fns/format'
import { createSlotSelector } from '../../reducers'

export const BookButton = ({ startTime, endTime}) => (
  <button>book { formatTime(startTime) } to { formatTime(endTime) }</button>
);

const formatTime = (time) => (
  format(time, 'H:mm')
);

const mapStateToProps = (state, { id }) => (
  createSlotSelector(id)
);

export default connect(mapStateToProps)(BookButton);
