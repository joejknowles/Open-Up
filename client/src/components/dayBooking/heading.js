import React from 'react';
import { connect } from 'react-redux';
import { selectedDateSelector } from '../../reducers';
import format from 'date-fns/format'
import { nextDay } from '../../actions/changeDay';

const dateTitle = (date) => (
  `Today ${format(date, 'D/M/YYYY')}`
);

export const Heading = ({ date, onNextDayClick }) => (
  <div className="Heading">
    <h3>{ dateTitle(date) }</h3>
    <a
      className="next-day"
      onClick={ onNextDayClick }>
      Next Day
    </a>
  </div>
);

export const mapStateToProps = (state) => ({
  date: selectedDateSelector(state)
});

const mapDispatchToProps = (dispatch) => ({ onNextDayClick: () => dispatch(nextDay) });

export default connect(mapStateToProps, mapDispatchToProps)(Heading);
