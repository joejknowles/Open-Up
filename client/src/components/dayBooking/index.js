import React, { Component } from 'react';
import '../../styles/App.css';
import { connect } from 'react-redux';
import { fetchSlots } from '../../actions';

import LoadingIndicator from '../loadingIndicator';
import Heading from './heading';
import List from './list';

export class DayBooking extends Component {
  componentDidMount() {
    this.props.fetchSlots();
  }

  render() {
    if (this.props.isLoading) return (<LoadingIndicator />);
    return (
      <div>
        <Heading />
        <List slots={ this.props.slots } />
      </div>
    );
  }
}

const mapStateToProps = (state) => ({
  slots: state.allSlots.map((id) => state.slotsById[id]),
  date: state.selectedDate,
  isLoading: state.isLoading
});

const mapDispatchToProps = (dispatch) => ({
  fetchSlots: fetchSlots(dispatch)
});

const mergeProps = ({ slots, isLoading, date }, dispatchProps) => ({
  slots, isLoading,
  fetchSlots: dispatchProps.fetchSlots(date)
});

export default connect(mapStateToProps, mapDispatchToProps, mergeProps)(DayBooking);
