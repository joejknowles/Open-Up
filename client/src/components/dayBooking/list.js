import React from 'react';
import '../../styles/DayBooking.css';
import { connect } from 'react-redux';
import { slotsSelector } from '../../reducers';
import Slot from './slot'

export const List = ({ slots = [] }) => (
  slots.length === 0 ?
  <p className="message">No slots available</p> :
  (<div className="List">
    { slots.map(( id ) => (
      <Slot key={ id } { ...{ id } } />
    )) }
  </div>)
);

const mapStateToProps = (state) => ({
  slots: slotsSelector(state)
});

const connectedList = connect(mapStateToProps)(List);

export default connectedList;
