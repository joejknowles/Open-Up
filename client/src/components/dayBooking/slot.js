import React from 'react';
import '../../styles/DayBooking.css';
import { connect } from 'react-redux';
import { createSlotSelector } from '../../reducers';
import BookButton from './bookButton';
import format from 'date-fns/format';

const formatTime = (time) => (
  format(time, 'HH:mm')
);

export const Slot = (props) => (
  <div className="Slot">
    {
      props.booking ?
        <UnavailableSlot { ...props } />
      :
        <BookButton { ...props }  />
    }
  </div>
);

export const UnavailableSlot = ({ startTime, endTime, booking }) => {
  const message = booking === "pending" ? '-' : `unavailable ${ formatTime(startTime) } to ${ formatTime(endTime) }`
  return (
    <div className="unavailableSlot slot-content">
      <span className="message">
        { message }
      </span>
    </div>
  );
}

const mapStateToProps = (state, { id }) => (
  createSlotSelector(id)
);

export default connect(mapStateToProps)(Slot);
