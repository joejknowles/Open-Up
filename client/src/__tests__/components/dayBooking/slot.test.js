import React from 'react';
import { shallow } from 'enzyme';
import { Slot, UnavailableSlot } from '../../../components/dayBooking/slot';
import { BookButton } from '../../../components/dayBooking/bookButton';
import ReactTestUtils from 'react-addons-test-utils';

const shallowRenderer = ReactTestUtils.createRenderer();

const startTime = new Date(null, null, null, 13);
const endTime = new Date(null, null, null, 14);
const times = { startTime, endTime };

it("renders a button when there's no booking", () => {
  const component = shallow(<Slot booking={ false } { ...times } />);
  expect(component.find('Connect(Component)').length).toBe(1);
});

it("renders unavailable message when there's a booking", () => {
  const component = shallow(<Slot booking={ true } { ...times } />);
  expect(component.find(UnavailableSlot).length).toBe(1);
});

it('renders Slot with booking as before', () => {
  const tree = shallowRenderer.render(
    <Slot booking={ true } { ...times } />
  );
  expect(tree).toMatchSnapshot();
});

it('renders Slot without booking as before', () => {
  const tree = shallowRenderer.render(
    <Slot booking={ false } { ...times } />
  );
  expect(tree).toMatchSnapshot();
});

it('renders UnavailableSlot as before', () => {
  const tree = shallowRenderer.render(
    <UnavailableSlot { ...times } />
  );
  expect(tree).toMatchSnapshot();
});

it("UnavailableSlot renders unavailable message", () => {
  const component = shallow(<UnavailableSlot { ...times } />);
  expect(component.find('span').text()).toBe('unavailable 13:00 to 14:00');
});
