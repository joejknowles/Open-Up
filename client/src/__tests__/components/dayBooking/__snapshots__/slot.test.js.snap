exports[`test renders Slot with booking as before 1`] = `
<div
  className="Slot">
  <Unknown
    booking={true}
    endTime={1899-12-31T14:00:00.000Z}
    startTime={1899-12-31T13:00:00.000Z} />
</div>
`;

exports[`test renders Slot without booking as before 1`] = `
<div
  className="Slot">
  <Connect(Component)
    booking={false}
    endTime={1899-12-31T14:00:00.000Z}
    startTime={1899-12-31T13:00:00.000Z} />
</div>
`;

exports[`test renders UnavailableSlot as before 1`] = `
<div
  className="unavailableSlot">
  <span
    className="message">
    unavailable 
    13:00
     to 
    14:00
  </span>
</div>
`;
